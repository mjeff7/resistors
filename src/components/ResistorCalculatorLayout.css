/*
 * Top level element holding the image and selectors.
 */

.imageAndSelectors {
  display: flex;
  height: 60vh;

  border-radius: 1rem;
  border: 1px solid lightgrey;
  padding: 1rem 2rem;
}

/*
 * Calculations display
 */

.calculatedResistanceDisplay {
  font-size: 1.5rem;
  font-family: 'Montserrat', sans-serif;

  margin: 1rem;
  padding: 0.5rem;

  border-radius: 1rem;
  border: 1px solid lightgrey;

  white-space: nowrap;
}

.calculatedResistanceDisplay > div {
  margin: 0.5rem;

  min-width: 14em;
}

.caption {
  font-size: 0.75rem;
  margin-top: 0.5rem;
  border-top: 1px solid black;
}

/*
 * Resistor image
 */

/* Make the resistor image respond to size of surroundings. It needs height or
 * width explicitly set on the parent in absolute terms, or an unbroken chain
 * of parents each setting those properties as percentages until the final
 * parent sets it in concrete terms. */
.resistorImageWrapper, .resistorImageWrapper > div {
  height: 100%;
  width: 100%;
}

svg {
  height: 100%;
  width: 100%;
}


/*
 * Selectors menus
 */

.selectors {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;

  padding-left: 1rem;
}

.selectors > select {
  font-size: 2rem;
  font-family: 'Maven Pro', sans-serif;
  padding: 0.5rem;
  border-radius: 1rem;
  font-variant: small-caps;

  min-width: 4em;
}

#bandA, #bandB, #bandC, #bandD, select {
  transition: 1s;
}
