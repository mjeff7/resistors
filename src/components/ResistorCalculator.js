// @flow

import { compose, defaultProps, withProps, withStateHandlers } from "recompose";

import {
  FIRST_DIGIT_COLORS,
  MULTIPLIER_COLORS,
  SECOND_DIGIT_COLORS,
  TOLERANCE_COLORS,
  type FirstDigitColor,
  type MultiplierColor,
  type SecondDigitColor,
  type ToleranceColor
} from "../calculator/colors";
import {
  type OhmValueCalculatorFromColors,
  attachToleranceBounds,
  calculateOhmValueFromColors
} from "../calculator/calculator";
import type { ToleranceValue } from "../calculator/values";
import { abbreviateValue } from "../utils";
import ResistorCalculatorLayout from "./ResistorCalculatorLayout";

export type BandColors = {
  bandAColor: FirstDigitColor,
  bandBColor: SecondDigitColor,
  bandCColor: MultiplierColor,
  bandDColor: ToleranceColor
};

export const useCalculator = (calculator: OhmValueCalculatorFromColors) =>
  compose(
    withProps(
      ({ bandAColor, bandBColor, bandCColor, bandDColor }: BandColors) => {
        // Explicitly destructure and return these properties so that Flow sees
        // that they are present as component properties.
        const { resistance, tolerance } = calculator(
          bandAColor,
          bandBColor,
          bandCColor,
          bandDColor
        );

        return { resistance, tolerance };
      }
    ),
    withProps((props: { resistance: number, tolerance: ToleranceValue }) => {
      const { minimum, maximum } = attachToleranceBounds(props);
      return { minimum, maximum };
    })
  );

export const attachStateHandlers = compose(
  withStateHandlers(({ bandAColor }) => ({ bandAColor }), {
    setBandAColor: () => (bandAColor: FirstDigitColor) => ({ bandAColor })
  }),
  withStateHandlers(({ bandBColor }) => ({ bandBColor }), {
    setBandBColor: () => (bandBColor: SecondDigitColor) => ({ bandBColor })
  }),
  withStateHandlers(({ bandCColor }) => ({ bandCColor }), {
    setBandCColor: () => (bandCColor: MultiplierColor) => ({ bandCColor })
  }),
  withStateHandlers(({ bandDColor }) => ({ bandDColor }), {
    setBandDColor: () => (bandDColor: ToleranceColor) => ({ bandDColor })
  })
);

export const formatResistanceValue = (value: number) => {
  const [smallValue, suffix] = abbreviateValue(value);

  return `${smallValue} ${suffix}Ω`;
};

export const formatTolerance = (value: ToleranceValue) => `± ${value * 100}%`;

const enhancer = compose(
  defaultProps({
    bandAColor: FIRST_DIGIT_COLORS[0],
    bandBColor: SECOND_DIGIT_COLORS[0],
    bandCColor: MULTIPLIER_COLORS[0],
    bandDColor: TOLERANCE_COLORS[0]
  }),
  attachStateHandlers,
  useCalculator(calculateOhmValueFromColors)
);

export default enhancer(ResistorCalculatorLayout);
