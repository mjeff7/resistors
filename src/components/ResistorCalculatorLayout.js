// @flow

import * as React from "react";

import type { BandColors } from "./ResistorCalculator";
import {
  FirstBandSelector,
  MultiplierBandSelector,
  SecondBandSelector,
  ToleranceBandSelector
} from "./bandSelectors";
import { interleave } from "../utils";
import ResistorImage from "./ResistorImage";

import "./ResistorCalculatorLayout.css";

type SpacerProps = {
  children: Array<React.Node>,
  spacing: Array<number>
};

const Spacer = ({ children, spacing = [], ...remainingProps }: SpacerProps) => {
  const spacers = spacing.map((spacerSize, spacerIndex) => (
    <div key={spacerIndex} style={{ flexGrow: spacerSize }} />
  ));

  const newChildren = interleave(spacers, children);

  return <div {...remainingProps}>{newChildren}</div>;
};

type Props = BandColors & {
  resistance: string,
  tolerance: string,
  minimum: string,
  maximum: string,
  setBandAColor: (*) => *,
  setBandBColor: (*) => *,
  setBandCColor: (*) => *,
  setBandDColor: (*) => *
};

export default ({
  resistance,
  tolerance,
  minimum,
  maximum,
  bandAColor,
  bandBColor,
  bandCColor,
  bandDColor,
  setBandAColor,
  setBandBColor,
  setBandCColor,
  setBandDColor
}: Props) => (
  <div>
    <div className="imageAndSelectors">
      <ResistorImage {...{ bandAColor, bandBColor, bandCColor, bandDColor }} />
      <Spacer className="selectors" spacing={[]}>
        <FirstBandSelector value={bandAColor} onSelectValue={setBandAColor} />
        <SecondBandSelector value={bandBColor} onSelectValue={setBandBColor} />
        <MultiplierBandSelector
          value={bandCColor}
          onSelectValue={setBandCColor}
        />
        <ToleranceBandSelector
          value={bandDColor}
          onSelectValue={setBandDColor}
        />
      </Spacer>
    </div>
    <div>
      Resistance value: {resistance} {tolerance}
    </div>
    <div>
      Range: {minimum} to {maximum}
    </div>
  </div>
);
