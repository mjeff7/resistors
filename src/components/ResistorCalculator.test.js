// @flow

import React from "react";

import { mount } from "enzyme";

import {
  FIRST_DIGIT_COLORS,
  MULTIPLIER_COLORS,
  SECOND_DIGIT_COLORS,
  TOLERANCE_COLORS
} from "../calculator/colors";
import {
  FirstBandSelector,
  MultiplierBandSelector,
  SecondBandSelector,
  ToleranceBandSelector
} from "./bandSelectors";
import {
  firstDigitFromColor,
  multiplierExponentFromColor,
  secondDigitFromColor,
  toleranceValueFromColor
} from "../calculator/values";
import ResistorCalculator, {
  attachStateHandlers,
  formatResistanceValue,
  formatTolerance,
  useCalculator
} from "./ResistorCalculator";
import ResistorCalculatorLayout from "./ResistorCalculatorLayout";

describe("useCalculator", () => {
  const setupTest = (a, b, c, d) => {
    const calculator = jest.fn((a, b, c, d) => ({
      resistance: 10 * a.length + b.length,
      tolerance: [0.2, 0.1, 0.05][(c.length + d.length) % 3]
    }));
    const SimpleComponent = () => <div />;
    const WithCalculator = useCalculator(calculator)(SimpleComponent);

    const props = mount(
      <WithCalculator
        bandAColor={a}
        bandBColor={b}
        bandCColor={c}
        bandDColor={d}
      />
    )
      .find(SimpleComponent)
      .props();

    return { props, calculator };
  };

  it("calls the calculator with passed parameters", () => {
    expect(
      setupTest("yellow", "green", "grey", "red").calculator
    ).lastCalledWith("yellow", "green", "grey", "red");

    expect(
      setupTest("orange", "black", "violet", "white").calculator
    ).lastCalledWith("orange", "black", "violet", "white");
  });

  it("supplies calculated properties as props", () => {
    expect(setupTest("yellow", "green", "grey", "red").props).toMatchObject({
      resistance: 65,
      tolerance: 0.1,
      minimum: 58.5,
      maximum: 71.5
    });

    expect(setupTest("blue", "red", "gold", "white").props).toMatchObject({
      resistance: 43,
      tolerance: 0.2,
      minimum: 34.4,
      maximum: 51.6
    });
  });
});

describe("numerical output", () => {
  let output1, output2;

  beforeAll(() => {
    output1 = mount(
      <ResistorCalculator
        bandAColor="blue"
        bandBColor="green"
        bandCColor="brown"
        bandDColor="violet"
      />
    ).text();

    output2 = mount(
      <ResistorCalculator
        bandAColor="white"
        bandBColor="grey"
        bandCColor="gold"
        bandDColor="silver"
      />
    ).text();
  });

  it("displays the resistance somewhere", () => {
    expect(output1).toContain("650 Ω");
    expect(output2).toContain("9.8 Ω");
  });

  it("displays the tolerance somewhere", () => {
    expect(output1).toContain("0.1%");
    expect(output2).toContain("10%");
  });

  it("displays the minimum somewhere", () => {
    expect(output1).toContain("649.35 Ω");
    expect(output2).toContain("8.82 Ω");
  });

  it("displays the maximum somewhere", () => {
    expect(output1).toContain("650.65 Ω");
    expect(output2).toContain("10.8 Ω");
  });
});

describe("state handlers effect color changes", () => {
  it("responds to changing colors", () => {
    const SimpleComponent = () => null;
    const Wrapped = attachStateHandlers(SimpleComponent);

    const component = mount(<Wrapped />);
    const getProps = () => component.find(SimpleComponent).props();
    const getProp = prop => getProps()[prop];

    const setAndTest = (handler, band, value) => {
      getProp(handler)(value);
      component.update();
      expect(getProp(band)).toBe(value);
    };

    setAndTest("setBandAColor", "bandAColor", "blue");
    setAndTest("setBandAColor", "bandAColor", "red");

    setAndTest("setBandBColor", "bandBColor", "white");
    setAndTest("setBandBColor", "bandBColor", "green");

    setAndTest("setBandCColor", "bandCColor", "pink");
    setAndTest("setBandCColor", "bandCColor", "silver");

    setAndTest("setBandDColor", "bandDColor", "violet");
    setAndTest("setBandDColor", "bandDColor", "yellow");
  });
});

describe("component interactivity", () => {
  it("responds to menu selections", () => {
    const component = mount(<ResistorCalculator />);
    const getProps = () => component.find(ResistorCalculatorLayout).props();
    const getProp = prop => getProps()[prop];

    const setAndTest = (Component, band, value) => {
      component
        .find(Component)
        .find("select")
        .simulate("change", { target: { value } });
      component.update();
      expect(getProp(band)).toBe(value);
    };

    setAndTest(FirstBandSelector, "bandAColor", "blue");
    setAndTest(FirstBandSelector, "bandAColor", "white");

    setAndTest(SecondBandSelector, "bandBColor", "red");
    setAndTest(SecondBandSelector, "bandBColor", "green");

    setAndTest(MultiplierBandSelector, "bandCColor", "pink");
    setAndTest(MultiplierBandSelector, "bandCColor", "yellow");

    setAndTest(ToleranceBandSelector, "bandDColor", "violet");
    setAndTest(ToleranceBandSelector, "bandDColor", "silver");
  });

  describe("displays current value as menu selection", () => {
    const setAndTest = (Component, band, value) => {
      expect(
        mount(<ResistorCalculator {...{ [band]: value }} />)
          .find(Component)
          .find("select")
          .props().value
      ).toBe(value);
    };

    it("first selector", () => {
      setAndTest(FirstBandSelector, "bandAColor", "blue");
      setAndTest(FirstBandSelector, "bandAColor", "white");
    });

    it("second selector", () => {
      setAndTest(SecondBandSelector, "bandBColor", "red");
      setAndTest(SecondBandSelector, "bandBColor", "green");
    });

    it("third selector", () => {
      setAndTest(MultiplierBandSelector, "bandCColor", "pink");
      setAndTest(MultiplierBandSelector, "bandCColor", "yellow");
    });

    it("fourth selector", () => {
      setAndTest(ToleranceBandSelector, "bandDColor", "violet");
      setAndTest(ToleranceBandSelector, "bandDColor", "silver");
    });
  });
});

describe("number formatting", () => {
  describe("resistance values", () => {
    const checkValue = value => {
      it(`${value}`, () => {
        expect(formatResistanceValue(value)).toMatchSnapshot();
      });
    };

    checkValue(0);

    checkValue(1);
    checkValue(1e-3);
    checkValue(1e3);
    checkValue(1e6);
    checkValue(1e9);

    checkValue(123);
    checkValue(9876543210);
  });

  describe("tolerance values", () => {
    const checkValue = value => {
      it(`${value}`, () => {
        expect(formatTolerance(value)).toMatchSnapshot();
      });
    };

    TOLERANCE_COLORS.forEach(color =>
      checkValue(toleranceValueFromColor(color))
    );
  });

  describe("values are guarded from rounding errors", () => {
    const checkValue = value => {
      it(`${value}`, () => {
        expect(formatResistanceValue(value).length).toBeLessThan(6);
      });
    };

    checkValue(10 * 1.2 * Math.pow(10, -1));

    const checkColorValueCombination = (firstDigit, secondDigit, multiplier) =>
      checkValue(
        Math.pow(10, multiplier) *
          (10 * firstDigit + secondDigit) *
          Math.pow(10, -multiplier)
      );

    FIRST_DIGIT_COLORS.forEach(bandAColor =>
      SECOND_DIGIT_COLORS.forEach(bandBColor =>
        MULTIPLIER_COLORS.forEach(bandCColor =>
          checkColorValueCombination(
            firstDigitFromColor(bandAColor),
            secondDigitFromColor(bandBColor),
            multiplierExponentFromColor(bandCColor)
          )
        )
      )
    );
  });
});
