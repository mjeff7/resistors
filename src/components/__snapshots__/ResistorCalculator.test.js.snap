// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`number formatting resistance values 0 1`] = `"0 Ω"`;

exports[`number formatting resistance values 0.001 1`] = `"1 mΩ"`;

exports[`number formatting resistance values 1 1`] = `"1 Ω"`;

exports[`number formatting resistance values 123 1`] = `"123 Ω"`;

exports[`number formatting resistance values 1000 1`] = `"1 kΩ"`;

exports[`number formatting resistance values 1000000 1`] = `"1 MΩ"`;

exports[`number formatting resistance values 1000000000 1`] = `"1 GΩ"`;

exports[`number formatting resistance values 9876543210 1`] = `"9.87654321 GΩ"`;

exports[`number formatting tolerance values 0.0005 1`] = `"± 0.05%"`;

exports[`number formatting tolerance values 0.001 1`] = `"± 0.1%"`;

exports[`number formatting tolerance values 0.005 1`] = `"± 0.5%"`;

exports[`number formatting tolerance values 0.0025 1`] = `"± 0.25%"`;

exports[`number formatting tolerance values 0.01 1`] = `"± 1%"`;

exports[`number formatting tolerance values 0.02 1`] = `"± 2%"`;

exports[`number formatting tolerance values 0.05 1`] = `"± 5%"`;

exports[`number formatting tolerance values 0.1 1`] = `"± 10%"`;

exports[`number formatting tolerance values 0.2 1`] = `"± 20%"`;
